index.js

```diff
@@ -102,6 +106,57 @@ function fastifyPostgres (fastify, options, next) {
     }
   }
 
+  if (!fastify.hasRequestDecorator('pg')) {
+    fastify.decorateRequest('pg', null)
+  }
+
+  fastify.addHook('onRoute', routeOptions => {
+    const useTransaction = routeOptions.pg && routeOptions.pg.transact
+
+    if (useTransaction) {
+      const preHandler = async (req, reply) => {
+        const client = await pool.connect()
+
+        if (name) {
+          if (client[name]) {
+            return next(new Error(`pg client '${name}' is a reserved keyword`))
+          } else if (fastify.pg[name]) {
+            return next(new Error(`request client '${name}' has already been registered`))
+          }
+
+          req.pg[name] = client
+        } else {
+          if (req.pg && req.pg.query) {
+            return next(new Error('request client has already been registered'))
+          } else {
+            req.pg = client
+          }
+        }
+
+        req.pg.query('BEGIN')
```

Comment:
if the plugin is registered with a name, you have to use it everywhere. this will not work. same in all other places where you're accessing req.pg.  all those accesses, in case a name is provided, should be req.pg[name]